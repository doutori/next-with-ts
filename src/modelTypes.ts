/**
 * Email
 *
 * @example "xxxx\u0040example.com"
 */
export type Email = string;

/**
 * Nickname
 *
 * @example "sid_jankovic_jeronimo_mongo"
 */
export type Nickname = string;

/**
 * Password
 */
export type Password = string;

/**
 * Access token (JWT)
 *
 * @example "eyJz93a...k4laUWw"
 */
export type AccessToken = string;

/**
 * Refresh token (JWT)
 *
 * @example "GEbRxBN...edjnXbL"
 */
export type RefreshToken = string;

/**
 * Account
 */
export interface Account {
    /**
     * Email
     */
    email: Email;

    /**
     * Nickname
     */
    nickname: Nickname;
}

export interface UpdateAccount extends Account {
    /**
     * Password as plain text
     */
    password?: Password;
}

export interface Token {
    /**
     * Access token (JWT)
     */
    access_token: AccessToken;

    /**
     * Refresh token (JWT)
     */
    refresh_token: RefreshToken;

    /**
     * Always 'Bearer'
     *
     * @example "Bearer"
     */
    type: string;
}

export interface RefreshTokenRequest {
    /**
     * Refresh token (JWT)
     */
    refresh_token: RefreshToken;
}

export interface SignupRequest extends Account {
    /**
     * Password as plain text
     */
    password: Password;
}

export interface SigninRequest {
    /**
     * Email
     */
    email: Email;

    /**
     * Password as plain text
     */
    password: Password;
}
