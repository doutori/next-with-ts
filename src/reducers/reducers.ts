export { account } from '~/reducers/account';
export { counter } from '~/reducers/counter';
