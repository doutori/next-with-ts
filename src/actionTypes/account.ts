import {
    INCREMENT,
    DECREMENT,
    UPDATE_COUNT,
    GET_ACCOUNT,
    UPDATE_ACCOUNT
} from '~/constant';
import { Account } from '~/modelTypes';

interface IncrementAction {
    type: typeof INCREMENT;
    payload: number;
}

interface DecrementAction {
    type: typeof DECREMENT;
    payload: number;
}

interface UpdateCountAction {
    type: typeof UPDATE_COUNT;
    payload: number;
}

export type CounterActionTypes =
    | IncrementAction
    | DecrementAction
    | UpdateCountAction;

interface GetAccountAction {
    type: typeof GET_ACCOUNT;
}

interface UpdateAccountAction {
    type: typeof UPDATE_ACCOUNT;
    payload: { data: Account };
}

export type AccountActionTypes = GetAccountAction | UpdateAccountAction;
