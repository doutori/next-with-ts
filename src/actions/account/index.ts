import { GET_ACCOUNT, UPDATE_ACCOUNT } from '~/constant';
import { AccountActionTypes } from '~/actionTypes';
import { Account } from '~/modelTypes';

export const getAccount = (): AccountActionTypes => ({
    type: GET_ACCOUNT
});

export const updateAccount = (account: Account): AccountActionTypes => ({
    type: UPDATE_ACCOUNT,
    payload: { data: account }
});
