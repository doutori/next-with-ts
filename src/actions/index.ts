import { INCREMENT, DECREMENT } from '../constant';
import { CounterActionTypes } from '../actionTypes';

export const increment = (number: number): CounterActionTypes => ({
    type: INCREMENT,
    payload: number
});

export const decrement = (number: number): CounterActionTypes => ({
    type: DECREMENT,
    payload: number
});
