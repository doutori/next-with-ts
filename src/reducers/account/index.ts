import { UPDATE_ACCOUNT } from '~/constant';
import { AccountActionTypes } from '~/actionTypes';
import { AccountState } from '~/stateTypes';

export const initialAccount: AccountState = {
    data: { email: '', nickname: '' }
};

export const account = (
    state = initialAccount,
    action: AccountActionTypes
): AccountState => {
    switch (action.type) {
        case UPDATE_ACCOUNT:
            return { ...state, data: action.payload.data };
        default:
            return state;
    }
};
