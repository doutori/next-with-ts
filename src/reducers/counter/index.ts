import { INCREMENT, DECREMENT, UPDATE_COUNT } from '~/constant';
import { CounterActionTypes } from '~/actionTypes';
import { CountState } from '~/stateTypes';

export const initialCounter: CountState = { count: 0 };

export const counter = (
    state = initialCounter,
    action: CounterActionTypes
): CountState => {
    switch (action.type) {
        case INCREMENT:
            return { ...state, count: state.count++ };

        case DECREMENT:
            return { ...state, count: state.count-- };

        case UPDATE_COUNT:
            return { ...state, count: action.payload };

        default:
            return state;
    }
};
