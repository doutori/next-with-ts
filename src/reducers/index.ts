import { combineReducers } from 'redux';
import * as reducers from './reducers';

export const rootReducer = combineReducers(reducers);

export type RootState = ReturnType<typeof rootReducer>;
