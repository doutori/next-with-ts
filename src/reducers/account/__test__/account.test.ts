import snapshotDiff from 'snapshot-diff';
import { accountActions } from '~/actions';

import { initialAccount, account } from '~/reducers/account';
import { Account } from '~/modelTypes';

const testingAccount: Account = {
    email: 'test@example.com',
    nickname: 'snapshot testing'
};

describe('Account reducer', () => {
    it('update account', () => {
        expect(
            snapshotDiff(
                initialAccount,
                account(
                    initialAccount,
                    accountActions.updateAccount(testingAccount)
                )
            )
        ).toMatchSnapshot();
    });
});
