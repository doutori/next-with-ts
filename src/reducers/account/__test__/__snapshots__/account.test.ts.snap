// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account reducer update account 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
    \\"data\\": Object {
-     \\"email\\": \\"\\",
-     \\"nickname\\": \\"\\",
+     \\"email\\": \\"test@example.com\\",
+     \\"nickname\\": \\"snapshot testing\\",
    },
  }"
`;
